include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,248 @@
+ * for each allocation. This is intended as a debug adaptor and shouldn't be used in
+ * performance sensitive code.
+ *
+ * @tparam Upstream Type of the upstream resource used for
+ * allocation/deallocation.
+ */
+template <typename Upstream>
+class tracking_resource_adaptor final : public device_memory_resource {
+ public:
+  // can be a std::shared_mutex once C++17 is adopted
+  using read_lock_t  = std::shared_lock<std::shared_timed_mutex>;
+  using write_lock_t = std::unique_lock<std::shared_timed_mutex>;
+
+  /**
+   * @brief Construct a new tracking resource adaptor using `upstream` to satisfy
+   * allocation requests and tracking active allocations.
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   *
+   * @param upstream The resource used for allocating/deallocating device memory
+   * @param capture_stacks If true, capture stacks for allocation calls
+   */
+  tracking_resource_adaptor(Upstream* upstream, bool capture_stacks = false)
+    : upstream_{upstream}, capture_stacks_{capture_stacks}
+  {
+    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
+  }
+
+  tracking_resource_adaptor()                                 = delete;
+  ~tracking_resource_adaptor()                                = default;
+  tracking_resource_adaptor(tracking_resource_adaptor const&) = delete;
+  tracking_resource_adaptor(tracking_resource_adaptor&&)      = default;
+  tracking_resource_adaptor& operator=(tracking_resource_adaptor const&) = delete;
+  tracking_resource_adaptor& operator=(tracking_resource_adaptor&&) = default;
+
+  /**
+   * @brief Return pointer to the upstream resource.
+   *
+   * @return Upstream* Pointer to the upstream resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @brief Checks whether the upstream resource supports streams.
+   *
+   * @return true The upstream resource supports streams
+   * @return false The upstream resource does not support streams.
+   */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+  /**
+   * @brief Print any outstanding allocations to debug log
+   *
+   */
+  void print_outstanding_allocations() const
+  {
+    read_lock_t lock(mtx);
+    std::ostringstream oss;
+    for (auto const& al : allocations) {
+      oss << "Allocation " << al.first << " of " << al.second.allocation_size << "bytes";
+      if (al.second.strace != nullptr) {
+        oss << " with callstack:" << std::endl << *al.second.strace;
+      }
+      oss << std::endl;
+    }
+
+    RMM_LOG_DEBUG("Outstanding Allocations: {}", oss.str());
```

Comment:
RMM_LOG_DEBUG is a noop unless the active level is <= SPDLOG_LEVEL_DEBUG, so no reason to build the string unless that is configured. This is why this function should be called `log_outstanding_allocations`. Also check if `allocations` is empty before doing any work, and simplifying the output text a bit

```suggestion
#if SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
    read_lock_t lock(mtx);
    if (not allocations.empty()) {
      std::ostringstream oss;
      for (auto const& al : allocations) {
        oss << al.first << " : " << al.second.allocation_size << " B";
        if (al.second.strace != nullptr) {
          oss << " : callstack:" << std::endl << *al.second.strace;
        }
        oss << std::endl;
      }

      RMM_LOG_DEBUG("Outstanding Allocations: {}", oss.str());
    }
#endif // SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
```

Reply:
spdlog requires a rebuild in order to see logging? That seems counter-productive.